### PR TITLE
Add Querying Functionality to OSB

### LINKED ISSUE #373: Add query testing to OpenSearch Benchmark Procedures
Related to #341 

Currently, with #364, we only support the indexing workload for OpenSearch benchmarks with k-NN indices. We should add another set of steps to test query performance, with measures of query latency, recall as well as some kind of memory metrics.

### PR DESCRIPTION
**Pull request by jmazanec15**
### Description
Adds ability to run query workload from a data set with OpenSearch Benchmark tool for k-NN workloads. Refactors some of the code to better share components across extensions. 

In addition, added unit tests for testing custom param sources.

For recall metrics, tracking issue here:  **https://github.com/opensearch-project/opensearch-benchmark/issues/199**. This will not be covered in this PR.
 
### Issues Resolved
#373 
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by travisbenedict** [comment]
I haven't used this functionality myself but this seems like it should work.

Is any of the data defined [here](https://github.com/jmazanec15/k-NN-1/blob/issue-373/benchmarks/osb/extensions/runners.py#L84) showing up in the results? Is it just `recall` that's missing?

If you haven't already you could try configuring OpenSearch Benchmark to write to an OpenSearch cluster. That would give you access to the full set of raw metrics.

File: benchmarks/osb/extensions/param_sources.py [hunk 1]
```diff
@@ -35,29 +61,143 @@ def __init__(self, workload, params, **kwargs):
...
+        Splits up the parameters source so that multiple clients can read data
+        from it.
+        Args:
+            partition_index: index of one particular partition
+            total_partitions: total number of partitions data set is split into
+
+        Returns:
+            The parameter source for this particular partion
+        """
+        if self.num_vectors % total_partitions != 0:
```
Comment (martin-gaievski) [inline benchmarks/osb/extensions/param_sources.py:83]: is this mean that the data set size must be divisible by the number of parallel clients? 
If so I think in next revision we need to relax this requirement and divide evenly except for last client that will have the remainder

**Review by martin-gaievski — COMMENTED** [review]
Few general questions:
- how we gonna handle OSB version updates, I think we use officially supported extension points, but just want to re-confirm that we minimize changes of breaking things on our end with upgrade to a new OSB version
- do you want to use multiple clients for queries in our benchmarks (say for k-NN release)? We probably need to come up with some formula to estimate number of clients based on cluster configuration.

File: benchmarks/osb/extensions/param_sources.py (hunk 1 above)
Comment (jmazanec15) [reply benchmarks/osb/extensions/param_sources.py:83]: Yes thats a good point. I can update this in a future PR. Will create an issue when PR is merged.

**Comment by jmazanec15** [comment]
> how we gonna handle OSB version updates, I think we use officially supported extension points, but just want to re-confirm that we minimize changes of breaking things on our end with upgrade to a new OSB version

@martin-gaievski Good question. I think it will most likely be addressed at a later date. Right now, we don't release the benchmarks as artifacts and we hard code dependency to OSB in requirements.txt.  I think eventually we will want to transfer things to https://github.com/opensearch-project/opensearch-benchmark-workloads/ and when we do that we can ensure version compatibility.

**Review by martin-gaievski — APPROVED** [review]

**Review by VijayanB — APPROVED** [review]

**Pull request merged**